### ISSUE TITLE
Programmatically added services are not initialized with lifecycle hooks

### ISSUE THREAD

**Issue by StephenEsser** [webdriverio/webdriverio#6503]
**Environment (please complete the following information):**
 - **WebdriverIO version:** 6.x.x
 - **Mode:** WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:** sync
 - **Node.js version:** 12.20.1
 - **NPM version:** 6.14.11
 - **Browser name and version:** N/A
 - **Platform name and version:** Mac OS Catalina

**Describe the bug**

When [running the test runner programmatically](https://webdriver.io/docs/clioptions#run-the-test-runner-programmatically), services added to the Launcher via `opts` are not initialized with the correct lifecycle hooks, but are merged into the services array.

**To Reproduce**

Add a service programmatically and observe the lifecycle hooks are never called. (before, beforeSession).

```
const WDIOService = require('../services/wdio-service');

const opts = { services: [WDIOService] };

const wdio = new Launcher('/path/to/my/wdio.conf.js', opts);

wdio.run().then((code) => {
    process.exit(code)
}, (error) => {
    console.error('Launcher failed to start the test', error.stacktrace)
    process.exit(1)
})
```

**Expected behavior**

Services added via the launcher programmatically should initialize and bind to the expected lifecycle hooks. 

Perhaps needing to be called with [addService](https://github.com/webdriverio/webdriverio/blob/v6.12.1/packages/wdio-config/src/lib/ConfigParser.ts#L59)

I can create an example project if necessary.

**Additional Details** 

We have a custom test runner that loops through test permutations. Each permutation needs to pass in information to the services prior to starting the launcher. Environment variables [do not work](https://github.com/webdriverio/webdriverio/issues/6411) here because the values change between runs and the wdio.conf is cached by node on the initial run.

Example:

Say we have four locales we want to test. The test runner can accept all of the locals and initiate a launcher for each local.

```js
// This example was free-handed so this is more or less reduced pseudo code. 

const locales = ['en', 'fr', 'de', 'bl'];

for(let index = 0; index < locales.length; index += 1) {
  // These service options need to change between runs and launchers.
  const opts = { services: [WDIOService, { locale: locales[index] }] };
  
  const launcher = new Launcher('/path/to/my/wdio.conf.js', opts);
  
  launcher.run();
} 
```

**Comment by christian-bromann** [comment]
@StephenEsser thanks for filing the issue. Unfortunately this is impossible to do as WebdriverIO uses worker processes and there is just no chance to reliably pass in a function object into a child process. While there are workarounds like message passing it would add tremendous complexity for something that offers too little value imo. For your case I suggest to pass in a local via environment variables, e.g.:

```js
// wdio.conf.js
export const config = {
  // ...
  services: [WDIOService, { locale: process.env.TEST_LOCAL }]
  // ...
}
```
and then use a package like [`npm-run-all`](https://www.npmjs.com/package/npm-run-all) to trigger your executions in parallel:

```
"scripts": {
  "test": "run-p test:locale:*",
  "test:locale:en": "TEST_LOCAL=en wdio run /path/to/my/wdio.conf.js",
  "test:locale:fr": "TEST_LOCAL=en wdio run /path/to/my/wdio.conf.js",
  "test:locale:de": "TEST_LOCAL=en wdio run /path/to/my/wdio.conf.js",
  "test:locale:bl": "TEST_LOCAL=en wdio run /path/to/my/wdio.conf.js",
```

I think this can be also achieved programmatically.

I will close this issue due to the reasons I mentioned above (it's to complex to implement while providing too little value). Happy to continue the conversation.

**Comment by StephenEsser** [comment]
Hi @christian-bromann, thanks for taking the time to look at this. Unfortunately environment variables and test scripts do not scale well for our use-case. Our test permutations also account for form factor and theme. A full permutation set (worst case example) would require ~65+ unique test scripts.

~ 13 locales [de en-AU en-CA en-GB en-US en es fr nl nl-BE pt sv sv-SE]
~ 5 form factors [tiny small medium large enormous]
~ 3 themes [default, lowlight, custom]

```
"scripts": {
  "test:locale:en:tiny:default-theme": "TEST_LOCAL=en FORM_FACTOR=tiny THEME=default-theme wdio run /path/to/my/wdio.conf.js",
  "test:locale:fr:tiny:default-theme": "TEST_LOCAL=fr FORM_FACTOR=tiny THEME=default-theme wdio run /path/to/my/wdio.conf.js",
  "test:locale:de:tiny:default-theme": "TEST_LOCAL=de FORM_FACTOR=tiny THEME=default-theme wdio run /path/to/my/wdio.conf.js",
  "test:locale:bl:tiny:default-theme": "TEST_LOCAL=bl FORM_FACTOR=tiny THEME=default-theme wdio run /path/to/my/wdio.conf.js",
  "test:locale:en:small:default-theme": "TEST_LOCAL=en FORM_FACTOR=small THEME=default-theme wdio run /path/to/my/wdio.conf.js",
  "test:locale:fr:small:default-theme": "TEST_LOCAL=fr FORM_FACTOR=small THEME=default-theme wdio run /path/to/my/wdio.conf.js",
  "test:locale:de:small:default-theme": "TEST_LOCAL=de FORM_FACTOR=small THEME=default-theme wdio run /path/to/my/wdio.conf.js",
  "test:locale:bl:small:default-theme": "TEST_LOCAL=bl FORM_FACTOR=small THEME=default-theme wdio run /path/to/my/wdio.conf.js",
  // ...
  ```

On CI we can dynamically flex using environment variables but locally we'd like to run these using a single command rather than managing so many test scripts manually or at the very least invoke the test runner with the unique set permutation.

> I think this can be also achieved programmatically.

Do you have an example in mind of invoking these programmatically? Are you thinking of something like a bash script invoked from node or a package.json script that invokes a bash script?

Locally our test run is invoked like this. `terra wdio` is a custom cli interface for invoking a test runner programmatically 

```
"scripts": {
    "wdio": "terra wdio --locales de en-AU --formFactors tiny large --theme default lowlight"
 }
```

**Comment by christian-bromann** [comment]
> Do you have an example in mind of invoking these programmatically? Are you thinking of something like a bash script invoked from node or a package.json script that invokes a bash script?

Yeah, you can take your Launcher example from above and set the environment variables accordingly, e.g.

```js
const locales = ['en', 'fr', 'de', 'bl'];

for(let index = 0; index < locales.length; index += 1) {
  process.env.TEST_LOCAL = locales[index]
  const launcher = new Launcher('/path/to/my/wdio.conf.js');
  launcher.run();
} 
```

**Comment by StephenEsser** [comment]
This was an approach we tried previously and it does not work. Node caches the wdio.conf file the first time it is read and never re-evaluates the file. So all envs will be set to the envs of the first run and never updated on subsequent runs

This issue has an example project attached for reference. 

https://github.com/webdriverio/webdriverio/issues/6411

**Comment by christian-bromann** [comment]
Can you set

```js
runnerEnv: {
  LOCALE: process.env.LOCALE
}
```

in your wdio config and see if it propagates?

**Comment by StephenEsser** [comment]
I gave this a shot and it looks like it locked the process.env down to the first initial option and didn't update. Do these changes match what you were requesting?

https://github.com/StephenEsser/wdio-service-cache-issue/compare/runner-env

Invoked locally by running:

`node ./run-launchers.js`

One option that resolves the issue is clearing the conf file cache in between programatic launcher options:

```js
 delete require.cache[require.resolve('./wdio.conf')];  // <- Whatever path is provided to the launcher
 ```
 
 The forces node to re-evaluate that file when being required on subsequent runs. It feels a little hacky though
